First, let us load the training and testing data from the adult dataset.

In [1]:
import numpy as np

X_train = np.loadtxt("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
                        usecols=(0, 4, 10, 11, 12), delimiter=", ")

y_train = np.loadtxt("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
                        usecols=14, dtype=str, delimiter=", ")

X_test = np.loadtxt("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test",
                        usecols=(0, 4, 10, 11, 12), delimiter=", ", skiprows=1)

y_test = np.loadtxt("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test",
                        usecols=14, dtype=str, delimiter=", ", skiprows=1)
y_test = np.array([a[:-1] for a in y_test])

## Naive Bayes with No Privacy
To begin, let us first train a regular (non-private) naive Bayes classifier, and test its accuracy.

In [2]:
from sklearn.naive_bayes import GaussianNB
nonprivate_clf = GaussianNB()
nonprivate_clf.fit(X_train, y_train)

from sklearn.metrics import accuracy_score

print("Non-private test accuracy: %.2f%%" % 
     (accuracy_score(y_test, nonprivate_clf.predict(X_test)) * 100))

Non-private test accuracy: 79.64%


## Differentially Private Naive Bayes Classification
First, install IBM Differential Privacy Library.

In [3]:
!pip install diffprivlib

     |████████████████████████████████| 71kB 5.4MB/s 
  Created wheel for diffprivlib: filename=diffprivlib-0.3.0-cp36-none-any.whl size=138999 sha256=22185221619386dc839e3b2d7b63bf732ac77bfa51155d771a4c695e64ed4080
  Stored in directory: /root/.cache/pip/wheels/64/68/62/617183f73d3feceab2c9d4081714a27bc11be5bb3f10f59b8a
Successfully built diffprivlib


Using the models.GaussianNB module of diffprivlib, we can train a naive Bayes classifier while satisfying differential privacy. If we don't specify any parameters, the model defaults to epsilon = 1.00.

In [4]:
import diffprivlib.models as dp
dp_clf = dp.GaussianNB()

dp_clf.fit(X_train, y_train)

print("Differentially private test accuracy (epsilon=%.2f): %.2f%%" % 
      (dp_clf.epsilon, accuracy_score(y_test, dp_clf.predict(X_test)) * 100))

Differentially private test accuracy (epsilon=1.00): 79.98%


/usr/local/lib/python3.6/dist-packages/diffprivlib/models/naive_bayes.py:102: PrivacyLeakWarning: Bounds have not been specified and will be calculated on the data provided. This will result in additional privacy leakage. To ensure differential privacy and no additional privacy leakage, specify bounds for each dimension.
  "privacy leakage, specify bounds for each dimension.", PrivacyLeakWarning)


As we can see from the output accuracies above, the regular (non-private) Naïve Bayes classifier could produce an accuracy of 79.64%, while setting epsilon=1.00, the differentially private Naïve Bayes classifier could achieve an accuracy of 78.59%. If we use a smaller epsilon, it usually leads to better privacy protection while less accuracy. For instance, if we set epsilon=0.01:

In [5]:
import diffprivlib.models as dp
dp_clf = dp.GaussianNB(epsilon=float("0.01"))
dp_clf.fit(X_train, y_train)

print("Differentially private test accuracy (epsilon=%.2f): %.2f%%" % 
      (dp_clf.epsilon, accuracy_score(y_test, dp_clf.predict(X_test)) * 100))

Differentially private test accuracy (epsilon=0.01): 76.91%


/usr/local/lib/python3.6/dist-packages/diffprivlib/models/naive_bayes.py:102: PrivacyLeakWarning: Bounds have not been specified and will be calculated on the data provided. This will result in additional privacy leakage. To ensure differential privacy and no additional privacy leakage, specify bounds for each dimension.
  "privacy leakage, specify bounds for each dimension.", PrivacyLeakWarning)
